### PR TITLE
Improve code generation for homogenous tuples of unknown length

### PR DESCRIPTION
**Pull request by Keno**
E.g.:
```
mutable struct foo
    x::Int
end
mutable struct bar
    tup::NTuple{N, foo} where N
end
function f(b::bar)
    sum = 0
    for i = 1:2
        @inbounds sum += b.tup[i].x
    end
    sum
end
global x = bar((foo(1),foo(2)))
@benchmark f(x::bar)
```

Before:
```
julia> @benchmark f(x::bar)
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     27.124 ns (0.00% GC)
  median time:      28.771 ns (0.00% GC)
  mean time:        29.864 ns (0.00% GC)
  maximum time:     116.762 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     995
```

After:
```
BenchmarkTools.Trial:
  memory estimate:  0 bytes
  allocs estimate:  0
  --------------
  minimum time:     10.531 ns (0.00% GC)
  median time:      10.559 ns (0.00% GC)
  mean time:        10.610 ns (0.00% GC)
  maximum time:     122.751 ns (0.00% GC)
  --------------
  samples:          10000
  evals/sample:     999
  time tolerance:   5.00%
  memory tolerance: 1.00%
```

### CODE REVIEW

File: src/cgutils.cpp [hunk 1]
```diff
@@ -1126,7 +1143,7 @@ static unsigned julia_alignment(Value* /*ptr*/, jl_value_t *jltype, unsigned ali
 static Value *emit_unbox(Type *to, const jl_cgval_t &x, jl_value_t *jt, Value* dest = NULL, bool volatile_store = false);
 
 static jl_cgval_t typed_load(Value *ptr, Value *idx_0based, jl_value_t *jltype,
-                             jl_codectx_t *ctx, MDNode *tbaa, unsigned alignment = 0)
+                             jl_codectx_t *ctx, MDNode *tbaa, bool maybe_null_if_boxed = true, unsigned alignment = 0)
```
Comment (yuyichao) [inline src/cgutils.cpp]: I think there are two users `emit_getfield_knownidx` and `emit_pointerref` that needs to be updated. (Or just add the argument at the end?)

**Review by yuyichao — DISMISSED** [review]

File: src/cgutils.cpp (hunk 1 above)
Comment (Keno) [inline src/cgutils.cpp]: Quite right of course. I backported this change from another branch where that function has a couple extra arguments. Fixed.

**Comment by ararslan** [comment]
@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21263#issuecomment-291293596) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/017dc389a5b6f754cf93e1948d2d10918dd80ea8/40c4860_vs_daefda4/report.md). cc @jrevels

File: src/cgutils.cpp [hunk 2]
```diff
@@ -560,14 +560,23 @@ static bool is_datatype_all_pointers(jl_datatype_t *dt)
...
-        if (!jl_is_leaf_type(t0))
+        if (!jl_is_leaf_type(t0)) {
+            if (allow_va && jl_is_vararg_type(t0) &&
+                  jl_is_leaf_type(jl_unwrap_vararg(t0)))
+                return true;
             return false;
+        }
         for(i=1; i < l; i++) {
+            if (allow_va && i == l - 1 && jl_is_vararg_type(jl_svecref(t,i))) {
+                if (!jl_types_equal(t0, jl_unwrap_vararg(jl_svecref(t,i))))
```
Comment (vtjnash) [inline src/cgutils.cpp]: this is not a well-formed type query (it can't be, since we've lost the UA bounds on `t` earlier)

File: src/codegen.cpp [hunk 3]
```diff
@@ -2967,6 +2967,33 @@ static bool emit_builtin_call(jl_cgval_t *ret, jl_value_t *f, jl_value_t **args,
...
+                    if (jl_is_vararg_type(jt))
+                        jt = jl_unwrap_vararg(jt);
+                    Value *vidx = emit_unbox(T_size, emit_expr(args[2], ctx), (jl_value_t*)jl_long_type);
+                    // This is not necessary for correctness, but allows to omit
+                    // the extra code for getting the length of the tuple
+                    if (!bounds_check_enabled(ctx)) {
+                        vidx = builder.CreateSub(vidx, ConstantInt::get(T_size, 1));
+                    } else {
+                        vidx = emit_bounds_check(strct, (jl_value_t*)stt, vidx,
+                            emit_datatype_nfields(emit_typeof(strct, ctx).V), ctx);
```
Comment (vtjnash) [inline src/codegen.cpp]: `emit_typeof_boxed`

**Comment by Keno** [comment]
A bunch of these results were also flakey in other logs. Let's try again to see if we have something real: @nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by Keno** [comment]
A bunch of these results were also flakey in other logs. Let's try again to see if we have something real:

@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by jrevels** [comment]
Good ol' UV_EAI_SYSTEM error stuck again. 

@nanosoldier `runbenchmarks(ALL, vs=":master")`

**Comment by nanosoldier** [comment]
[Your benchmark job](https://github.com/JuliaLang/julia/pull/21263#issuecomment-292301329) has completed - possible performance regressions were detected. A full report can be found [here](https://github.com/JuliaCI/BaseBenchmarkReports/blob/29618a797cd24e069f0509c7e080af18151f155a/165ec20_vs_d20c66d/report.md). cc @jrevels

File: src/cgutils.cpp (hunk 2 above)
Comment (Keno) [inline src/cgutils.cpp]: @JeffBezanson Since I know `t0` is a leaf type here, can I just use simple pointer equality?

Comment (JeffBezanson) [inline src/cgutils.cpp]: Yes.

File: src/codegen.cpp [hunk 4]
```diff
@@ -3010,6 +3010,33 @@ static bool emit_builtin_call(jl_cgval_t *ret, jl_value_t *f, jl_value_t **args,
                     }
                 }
             }
+        } else {
+            jl_value_t *utt = jl_unwrap_unionall((jl_value_t*)stt);
+            if (jl_is_tuple_type(utt) && is_tupletype_homogeneous(((jl_datatype_t*)utt)->types, true)) {
+                // For tuples, we can emit code even if we don't know the exact
+                // type (e.g. because we don't know the length). So long as all
+                // elements are the same and a leaf type.
```
Comment (tkelman) [inline src/codegen.cpp]: the same type, right? "elements are the same" sounds like it might mean value

Comment (Keno) [inline src/codegen.cpp]: Yes, it's worded a bit awkwardly, but it's `are the same [type] and a leaf type`. I can update the comment.

Comment (tkelman) [inline src/codegen.cpp]: that's better thanks

**Pull request merged**